creusot/src/backend/clone_map/elaborator.rs

```diff
@@ -1012,6 +1009,99 @@ fn fn_mut_hist_inv_term<'tcx>(
+    ctx: &Why3Generator<'tcx>,
+    typing_env: TypingEnv<'tcx>,
+    def_id: DefId,
+    subst: GenericArgsRef<'tcx>,
+) -> Option<Term<'tcx>> {
+    use rustc_type_ir::TyKind::*;
+    let int_ty = ctx.sig(def_id).output;
+    let arg = subst.type_at(0);
+    match arg.kind() {
+        Bool | Char | Int(_) | Uint(_) | Float(_) => {
+            Some(Term::int(int_ty, arg.primitive_size(ctx.tcx).bytes() as i128))
+        }
+        // "The types *const T, &T, Box<T>, Option<&T>, and Option<Box<T>> all have the same size. If T is Sized, all of those types have the same size as usize.
+        // The mutability of a pointer does not change its size. As such, &T and &mut T have the same size. Likewise for *const T and *mut T."
+        RawPtr(t, _) | Ref(_, t, _) if t.is_sized(ctx.tcx, typing_env) => {
+            Some(pointer_size(ctx.tcx, int_ty))
+        }
+        Adt(def, args) if is_pointer_adt(ctx.tcx, typing_env, def, args) => {
+            Some(pointer_size(ctx.tcx, int_ty))
+        }
+        Tuple(t) if t.len() == 0 => Some(Term::int(int_ty, 0)),
+        Adt(_, _) => None, // TODO
+        Array(t, n) => size_of_array(ctx, def_id, *t, n, int_ty),
+        _ => None,
+    }
+}
+
+/// `Box<T>`, `Option<&T>`, or `Option<Box<T>>` with `T: Sized`
+fn is_pointer_adt<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    typing_env: TypingEnv<'tcx>,
+    def: &ty::AdtDef<'tcx>,
+    args: GenericArgsRef<'tcx>,
+) -> bool {
+    let is_box = def.is_box() && args.type_at(0).is_sized(tcx, typing_env);
+    let is_option_ref =
+        || is_option(tcx, def) && is_ref_or_box_of_sized(tcx, typing_env, &args.type_at(0));
+    is_box || is_option_ref()
+}
+
+fn is_option(tcx: TyCtxt, def: &ty::AdtDef) -> bool {
+    tcx.lang_items().option_type().unwrap() == def.did()
+}
+
+fn is_ref_or_box_of_sized<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    typing_env: TypingEnv<'tcx>,
+    ty: &Ty<'tcx>,
+) -> bool {
+    use rustc_type_ir::TyKind::*;
+    match ty.kind() {
+        Ref(_, t, _) => t.is_sized(tcx, typing_env),
+        Adt(def, args) => def.is_box() && args.type_at(0).is_sized(tcx, typing_env),
+        _ => false,
+    }
+}
+
+fn pointer_size<'tcx>(tcx: TyCtxt<'tcx>, int_ty: Ty<'tcx>) -> Term<'tcx> {
+    Term::int(int_ty, tcx.data_layout().ptr_sized_integer().size().bytes() as i128)
+}
+
+fn size_of_array<'tcx>(
+    ctx: &Why3Generator<'tcx>,
+    def_id: DefId,
+    ty: Ty<'tcx>,
+    n: &Const<'tcx>,
+    int_ty: Ty<'tcx>,
+) -> Option<Term<'tcx>> {
+    let n = match n.kind() {
+        ConstKind::Value(_, ty::ValTree::Leaf(scalar)) => scalar.to_target_usize(ctx.tcx) as i128,
+        // TODO: ConstKind::Param
+        _ => return None,
+    };
+    if n == 0 {
+        Some(Term::int(int_ty, 0))
+    } else {
+        let subst = ty::GenericArgs::for_item(ctx.tcx, def_id, |_, _| ty.clone().into());
+        let item_size = Term {
+            ty: int_ty,
+            span: DUMMY_SP,
+            kind: TermKind::Call { id: def_id, subst, args: [].into() },
```

Comment:
Cannot you use `Term::call` instead? I'm trying to force us using these smart constructors, because they are so much less verbose.